### PR TITLE
Xiaow improve loading speed of wbs pages

### PR DESCRIPTION
**Pull request by wang9hu**
# Description
Screenshot 2023-06-22 175925

## Main changes explained:
1. Fix the freeze problem caused by `<ReactTooltip>`:  The reason why browser freeze is because each `<Task>` component has a `<ReactTooltip />` component which will create a `<div>` on the page, and when task number goes up, there will be hundreds of them.  This PR address the problem by running `ReactTooltip.rebuild()` on the parent component (`<WBSTasks>`) to represent all the tooltips of the children component (`<Task>`), so that only one `<div>` will be created;
2. Optimize the `allTasksReducer.js` logic: 

- Previously, app would send 5 requests for different levels of tasks, and the order of receiving responses is not consistent. Also, when making level 0 request, app will empty all previous cached tasks, which will cause tasks of other levels to be removed if their responses arrive earlier. This PR address this issue by adding a `fetchedData` buffer in the state, and the response data will have fixed position in the buffer based on their level. `sortByNum` function will combine and sort all the response data into `taskItems`.
- Because now the `sortByNum` function will handle the sorting for all tasks, for case like `case types.ADD_NEW_TASK:`, no need to find the index of mother task and insert at the specific position. Just add the new task into the `taskItems` and sort them again.
- Based on my observation, the `addTask` functionality now is not working properly, the created new task does not change its parent task property `hasChild` from default `false` to `true`. I did a temporary fix in the `sortByNum` function to create a `hasChildren` property representing the actual `hasChild` value, so that the tasks can have a folder icon/bold look on the page if they have any sub-tasks. This needs to be fixed in the future PR.
- I noticed there is a parameter called `mother` in the `case types.RECEIVE_TASKS:`, but I didn't find any code that use this parameter for `case types.RECEIVE_TASKS:`, except for the `allTasksReducer.test.js` test file so I commented out the old code for possible future reference.

3. Modify `allTasksReducer.test.js` test file:
- To cope with the newly added property in the `allTasksReducer.js` state, 
- I notice some code in previous version of `allTasksReducer.js` are just for passing the test file and does not have any real functional value. What I have seen so far in  are:
  - In `allTasksReducer.js`: 
     - `hasChildren` was created but never used except by `allTasksReducer.test.js` test files ; 
     - `taskId` property was not in the response data and it only showed up in the `allTasksReducer.test.js` test files;
     - `mother` parameter was never used outside of `allTasksReducer.test.js` test files; 
  -  In `allTasksReducer.test.js`;
     - `level: 0` was never used since all levels start from 1. 
- So I disabled some test cases and made some modification. 

4. Remove redundant sorting logic in `<WBSTasks>`

## How to test:
1. check into current branch
2. do `npm install` and `npm run start:local` to run this PR locally
3. Be sure to use Incognito mode or clear your cache before testing
4. log as admin user
5. go to dashboard → Other Links →  Projects →  WBS →  Pick any (try 'AAA USE THIS PROJECT' → 'Jae's WBS - USE THIS', which has 300+ tasks)
6. verify no more freezing or abnormal behavior of this page

## Note:
I noticed the filter functionality (unfold or fold tasks) is inconsistent upon refreshing or clicking the refresh button, but that is beyond the scope of this PR. Maybe this could be addressed in the future.

### CODE REVIEW

**Comment by EvianTan** [comment]
Be sure to clear your Chrome cache before testing.
Nice descriptions and explanation of changes above! Thank you.

**Review by EvianTan — DISMISSED** [review]

**Review by AuroraLovesWhales — DISMISSED** [review]
Hi @wang9hu. I have tested your PR and there is a significant improvement in loading speed. Great work!

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/44281916/7af7b993-720a-4acb-9c69-4dfdddd1451a

**Comment by wang9hu** [comment]
> Be sure to clear your Chrome cache before testing.
> Nice descriptions and explanation of changes above! Thank you.

Thank you for testing my PR! Could you please take a look at the changes I made to the test files? I think the changes won't mess other things up, but it is good to get some feedback from other perspectives.

**Comment by wang9hu** [comment]
> Hi @wang9hu. I have tested your PR and there is a significant improvement in loading speed. Great work!
> 
> https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/44281916/7af7b993-720a-4acb-9c69-4dfdddd1451a

Thank you! Could you also please give me some feedback on the changes in the test file?

**Comment by rachnapandya** [comment]
I had the same error earlier but now it works way better and faster.
Great work. @wang9hu 

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/53634437/af6a5832-3f5d-4d7f-b1f5-203b1ad9c6a3

**Comment by AnishPandita** [comment]
Hi @wang9hu I tested the PR and the loading time does seem reduced. Great Job.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/111321945/53d0914e-4007-4603-ac50-872ee05dd6b5

File: src/reducers/allTasksReducer.js [hunk 1]
```diff
@@ -15,38 +16,25 @@ const filterAndSort = (tasks, level) => {
...
-      task.num += '.0.0.0';
-    }
-    if (task.level === 2) {
-      task.num += '.0.0';
-    }
-    if (task.level === 3) {
-      task.num += '.0';
-    }
+    const numOfNums = task.num.split('.').length;
+    const num = task.num.concat('.0'.repeat(4 - numOfNums));
```
Comment (suigunfjc) [inline src/reducers/allTasksReducer.js:35]: Good to have a comment here

**Review by suigunfjc — APPROVED** [review]
Hi @wang9hu, I tested all features related to your PR and they all works as intended. Additionally, I have thoroughly reviewed your code changes, and I think they are clean and impressive. However, I would recommend adding some comments to certain line to enhance readability. Overall, excellent work!

**Comment by EdwinLauM** [comment]
Works as intended! Task loading is definitely faster! Good job!

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/42288987/76b316ce-f4d3-4001-87a9-739488e220f2

**Review by EdwinLauM — DISMISSED** [review]
Works as intended! Task loading is definitely faster! Good job!

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/42288987/76b316ce-f4d3-4001-87a9-739488e220f2

**Pull request merged**